nntrainer/src/neuralnet.cpp

```diff
@@ -438,92 +433,54 @@ int NeuralNetwork::setTrainConfig(std::vector<std::string> values) {
 
 int NeuralNetwork::init() {
   int status = ML_ERROR_NONE;
-  bool last = false;
   TensorDim previous_dim, def_init_dim;
 
+  status = checkValidation();
+  NN_RETURN_STATUS();
+
   /** Note: number of entries in layers will change. */
   for (unsigned int i = 0; i < layers.size(); ++i) {
-    if (i == layers.size() - 1)
-      last = true;
-
-    VERIFY_SET_DIMENSION();
-
-    switch (layers[i]->getType()) {
-    case LAYER_IN: {
-      status = layers[i]->initialize(last);
-      NN_RETURN_STATUS();
-    } break;
-    case LAYER_CONV2D: {
-      std::shared_ptr<Conv2DLayer> conv2d_layer =
-        std::static_pointer_cast<Conv2DLayer>(layers[i]);
-
-      status = layers[i]->setCost(cost);
-      NN_RETURN_STATUS();
-
-      status = layers[i]->initialize(last);
-      NN_RETURN_STATUS();
-
-      status = conv2d_layer->setOptimizer(opt);
-      NN_RETURN_STATUS();
-
-    } break;
-
-    case LAYER_POOLING2D: {
-      std::shared_ptr<Pooling2DLayer> pooling2d_layer =
-        std::static_pointer_cast<Pooling2DLayer>(layers[i]);
-
-      status = layers[i]->initialize(last);
-      NN_RETURN_STATUS();
-
-    } break;
-
-    case LAYER_FLATTEN: {
-      std::shared_ptr<FlattenLayer> flatten_layer =
-        std::static_pointer_cast<FlattenLayer>(layers[i]);
-
-      status = layers[i]->initialize(last);
-      NN_RETURN_STATUS();
-    } break;
-
-    case LAYER_FC: {
-      std::shared_ptr<FullyConnectedLayer> fc_layer =
-        std::static_pointer_cast<FullyConnectedLayer>(layers[i]);
-
-      status = layers[i]->setCost(cost);
-      NN_RETURN_STATUS();
-
-      status = layers[i]->initialize(last);
-      NN_RETURN_STATUS();
+    bool last = i == layers.size() - 1;
+    bool first = i == 0;
+    Layer &l = *layers[i];
+
+    if (!first) {
+      /// @fixme 359
+      if (def_init_dim == l.getInputDimension()) {
+        l.setInputDimension(previous_dim);
+      } else if (previous_dim != l.getInputDimension()) {
+        ml_loge("Dimension mismatch between layers.");
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+    }
 
-      status = fc_layer->setOptimizer(opt);
-      NN_RETURN_STATUS();
+    status = layers[i]->initialize(last);
 
-    } break;
+    switch (l.getType()) {
     case LAYER_BN:
-      status = layers[i]->initialize(last);
-      NN_RETURN_STATUS();
-
-      status = layers[i]->setOptimizer(opt);
-      NN_RETURN_STATUS();
-
-      if (i == 0) {
+      if (first) {
         ml_loge("Batch normalization layer cannot be first layer of network");
         status = ML_ERROR_INVALID_PARAMETER;
         NN_RETURN_STATUS();
       }
-      layers[i - 1]->setBNfollow(true);
+      /// no break intended
+    case LAYER_CONV2D:
```

Comment:
can you add something like fallthrough is intended?

Reply:
Yup. Fixed :)